### PR TITLE
CRM-17267 Fix change event selections impossible for sold out options

### PR DESCRIPTION
**Pull request by JKingsnorth**
See https://issues.civicrm.org/jira/browse/CRM-17267

### CODE REVIEW

**Comment by JKingsnorth** [comment]
Just need to double check this use of is_full:
https://github.com/JKingsnorth/civicrm-core/blob/f8f517faebfcfac374d2908e867f572cba74c073/CRM/Event/Form/Registration/Register.php#L228

**Comment by JKingsnorth** [comment]
Yes, this is ready for QA

**Comment by JKingsnorth** [comment]
Jenkins, would you be so kind as to retest this please?

**Comment by eileenmcnaughton** [comment]
@JKingsnorth I tried to replicate this & failed - the screenshot is without - you can see I have the option of deselecting

screenshot 2018-04-04 17 59 17

I DID hit #11934

**Comment by JKingsnorth** [comment]
Clarified the steps to recreate on JIRA. Try it with a 'checkboxes' and a 'required' field.

**Comment by eileenmcnaughton** [comment]
Ok I can see it now

BEFORE

screenshot 2018-04-04 20 33 50

AFTER
screenshot 2018-04-04 20 34 53


What bothers me about this is that 
a) we are allowing back office user to select a sold out option - I can see a case for and against this but
b) we not notifiying them in any way the difference between them submitting the form with the sold out option STILL selected (ie. sold out total NOT exceeded) and the sold out option NEWLY selected (ie if they submit the form the sold out total WILL be exceeded)

I feel like we need to mark this pretty clearly in the UI if we are going to make this change. Perhaps a confirm popup on submit (although my recollection is we don't have a good easy way to do this)  if sold out is exceeded or js to add some red text beside the option that will exceed saying 'if you submit this the event will be oversubscribed'



ping @agh1 @Stoob @jusfreeman @colemanw  if anyone has thoughts

**Comment by agh1** [comment]
@JKingsnorth that makes sense to have the option only be available if it's already selected.

If that's not feasible, or if you prefer, one idea might be to have a checkbox called "Show sold-out options".  That might act a little like the glass you need to break before pulling the fire alarm.

File: CRM/Event/Form/Registration/Register.php [hunk 1]
```diff
@@ -678,7 +678,12 @@ public static function formatFieldsForOptionFull(&$form) {
       }
     }
 
+    // Do not freeze full options on the back-end registration forms
     $className = CRM_Utils_System::getClassName($form);
+    $overrideIsFull = FALSE;
+    if ($className == 'CRM_Event_Form_Participant' || $className == 'CRM_Event_Form_ParticipantFeeSelection') {
+      $overrideIsFull = TRUE;
+    }
```
Comment (monishdeb) [inline CRM/Event/Form/Registration/Register.php]: @JKingsnorth lets change it 1 line:
```
$overrideIsFull = (in_array($className, ['CRM_Event_Form_Participant', 'CRM_Event_Form_ParticipantFeeSelection']));
```

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Register.php]: @monishdeb you mean change that one line in this PR?

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Register.php]: oh I see - quicker way to say it

File: CRM/Event/Form/Registration/Register.php [hunk 2]
```diff
@@ -729,16 +735,12 @@ public static function formatFieldsForOptionFull(&$form) {
             $isFull = TRUE;
           }
         }
-        $option['is_full'] = $isFull;
+        // 'is_full' says whether we should freeze the option in \CRM_Price_BAO_PriceField::freezeIfEnabled
+        $option['is_full'] = $overrideIsFull ? FALSE : $isFull;
```
Comment (monishdeb) [inline CRM/Event/Form/Registration/Register.php]: Oh we can further reduce it too:
```$option['is_full'] = (in_array($className, ['CRM_Event_Form_Participant', 'CRM_Event_Form_ParticipantFeeSelection'])) ? FALSE : $isFull;```

File: CRM/Event/Form/Registration/Register.php (hunk 1 above)
Comment (monishdeb) [reply CRM/Event/Form/Registration/Register.php]: Yes, but this is no longer required if we move this check below at L739 :)

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Register.php]: @monishdeb I think if ($form->isBackOffice) would work :-)

**Comment by eileenmcnaughton** [comment]
@monishdeb so the issue here is that it unlocks the full options when not selected too I think - but if we could resolve a new PR & get it merged....

File: CRM/Event/Form/Registration/Register.php (hunk 1 above)
Comment (monishdeb) [reply CRM/Event/Form/Registration/Register.php]: Umm nope as ```CRM_Event_Form_ParticipantFeeSelection``` doesn't inherit that property. Unless we should add the property ```isBackOffice``` to ```CRM_Core_Form``` and let the child form decide wether they like to be in front or as a silent worker

Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Register.php]: We can just add it to CRM_Event_Form_ParticipantFeeSelection I guess - if it IS always backoffice?

**Comment by eileenmcnaughton** [comment]
@monishdeb I'm inclined to close this in favour of a much simpler PR

```
     $defaultPricefieldIds = array();
     if (!empty($form->_values['line_items'])) {
       foreach ($form->_values['line_items'] as $lineItem) {
-        $defaultPricefieldIds[] = $lineItem['price_field_value_id'];
+        if ($lineItem['qty'] > 0) {
+          $defaultPricefieldIds[] = $lineItem['price_field_value_id'];
+        }
       }
     }
     if (!$priceSetId ||
@@ -724,7 +726,7 @@ class CRM_Event_Form_Registration_Register extends CRM_Event_Form_Registration {
         ) {
           $isFull = TRUE;
           $optionFullIds[$optId] = $optId;
-          if ($field['html_type'] != 'Select') {
+          if ($field['html_type'] != 'Select' && $field['html_type'] !== 'CheckBox') {
             if (in_array($optId, $defaultPricefieldIds)) {
               $optionFullTotalAmount += CRM_Utils_Array::value('amount', $option);
             }

```

What do you think?

**Comment by monishdeb** [comment]
@eileenmcnaughton that patch makes sense to me. Can you create a new PR with before/after screenshot. Will then close it in favor of that PR.

**Comment by JKingsnorth** [comment]
I'm happy to review and test the new PR

**Comment by eileenmcnaughton** [comment]
@JKingsnorth @monishdeb done https://github.com/civicrm/civicrm-core/pull/13096

I'll close this in favour of that

**Pull request closed without merging**